actions/sync-calendar/gcal/gcal.go

```diff
@@ -0,0 +1,131 @@
+// Package gcal handles Google Calendar integration.
+package gcal
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/spinnaker/rotation-scheduler/schedule"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/calendar/v3"
+	"google.golang.org/api/option"
+)
+
+const (
+	UserAgent  = "github.com/spinnaker/rotation-scheduler"
+	DateFormat = "2006-01-02"
+
+	defaultEndpoint = "https://www.googleapis.com/calendar/v3/"
+)
+
+// GCal wraps the Google Calendar service.
+type GCal struct {
+	CalendarID string
+	svc        *calendar.Service
+}
+
+// NewGCal wraps the calendar specified using the client. CalendarID should be a user's primary calendar (not a shared,
+// calendar), because the Clear() method only works on primary calendars.
+func NewGCal(calendarID string, client *http.Client) (*GCal, error) {
+	if calendarID == "" {
+		return nil, fmt.Errorf("calendar ID cannot be empty")
+	}
+
+	svc, err := calendar.NewService(context.Background(),
+		option.WithHTTPClient(client),
+		option.WithUserAgent(UserAgent),
+		option.WithEndpoint(defaultEndpoint))
+	if err != nil {
+		return nil, fmt.Errorf("unable to create Calendar service: %v", err)
+	}
+
+	return &GCal{
+		CalendarID: calendarID,
+		svc:        svc,
+	}, nil
+}
+
+// Client creates an authenticated HTTP Client from the specified service account credentials. The service account
+// must have a Domain-wide Delegation to control user calendars. See
+// https://developers.google.com/admin-sdk/directory/v1/guides/delegation
+func Client(calendarID, jsonCredentialsPath string) (*http.Client, error) {
+	key, err := ioutil.ReadFile(jsonCredentialsPath)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read JSON credential file: %v", err)
+	}
+	jwtConfig, err := google.JWTConfigFromJSON(key, calendar.CalendarScope)
+	if err != nil {
+		return nil, fmt.Errorf("unable to generate config from JSON credential: %v", err)
+	}
+	// Since apparently service accounts don't have any associated quotas in GSuite,
+	// we must supply a user to charge quota against, and I think they need to have
+	// admin permission on the G Suite account to work.
+	jwtConfig.Subject = calendarID
+
+	return jwtConfig.Client(context.Background()), nil
+}
+
+// Clears all events on the calendar and replaces them with events from Schedule sched.
+func (g *GCal) Schedule(sched *schedule.Schedule) error {
+	if err := sched.Validate(); err != nil {
+		return fmt.Errorf("schedule is invalid: %v", err)
+	}
```

Comment:
This still feels so weird to me... In Java you would have a `Schedule` object that could only be mutated by specific methods such that it was _impossible_ to have an object in an invalid state... That's why we usually wrap proto objects in Java too.

Now that you got rid of the proto object, you're back to essentially just passing around raw YAML files, which seems like it has the same issues... you have to constantly wonder if your `Schedule` struct is in a valid state or not. I dunno, maybe this is how things are done in Go?

Reply:
Public fields are a common thing in Go - for better or worse. Even the generated Go libraries from protos make their field public, which just means anything with access to that proto can set it. 

> you're back to essentially just passing around raw YAML files

Not quite - we're passing around `time.Time` values, which are much better to work with since we don't have to parse the human-readable `string` values, and thus deal with the `error`s that come with that (and we get the benefit of the rest of the `time` package for manipulating and printing those values)